### PR TITLE
[DOM] NaiveASTFlattener should be sync with ASTFlattener of the JDT.UI

### LINKED ISSUE #3261: [DOM] NaiveASTFlattener should be sync with ASTFlattener of the JDT.UI
The [unnamed variables and pattern](https://openjdk.org/jeps/456)'s DOM and ASTRewrite implemented through #2623. Following items should be added along with that. 

- The Javadoc of both APIs i.e. TypePattern.getPatternVariable() and TypePattern.getPatternVariable2() is same and not clear regarding the difference.
- The methods in JDT UI's ASTFlattener are kept in sync with NaiveASTFlattener(TypePattern). Hence, the change should be done in NaiveASTFlattener as well and then synced to ASTFlattener.

The Problem has reported here: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1764#issuecomment-2458775954

### PR DESCRIPTION
**Pull request by subyssurendran666**
## What it does
The [unnamed variables and pattern](https://openjdk.org/jeps/456)'s DOM and ASTRewrite implemented through https://github.com/eclipse-jdt/eclipse.jdt.core/issues/2623. Following items should be added along with that.

- The Javadoc of both APIs i.e. TypePattern.getPatternVariable() and TypePattern.getPatternVariable2() is same and not clear regarding the difference.
- The methods in JDT UI's ASTFlattener are kept in sync with NaiveASTFlattener(TypePattern). Hence, the change should be done in NaiveASTFlattener as well and then synced to ASTFlattener.

The Problem has reported here: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1764#issuecomment-2458775954

## How to test
This PR closes #3261 

## Author checklist

- [x] I have thoroughly tested my changes
- [x] The change is following the [coding conventions](https://wiki.eclipse.org/Platform/How_to_Contribute#Coding_Conventions)
- [x] I have signed the [Eclipse Contributor Agreement (ECA)](https://www.eclipse.org/legal/ECA.php)

### CODE REVIEW

**Comment by subyssurendran666** [comment]
@mpalat could you please review this PR and share your insights on this ?

**Comment by mpalat** [comment]
> @mpalat could you please review this PR and share your insights on this ?

Will take a look. Meanwhile, please have a look and follow the process at https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/Eclipse_API_Central_Deprecation_Policy.md when you are looking at deprecating APIs

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java [hunk 1]
```diff
@@ -247,12 +248,12 @@ public SingleVariableDeclaration getPatternVariable() {
...
 	 *
 	 * @return the pattern variable
-	 * @exception UnsupportedOperationException if this operation is used other than JLS19
+	 * @exception UnsupportedOperationException if this operation is used other than JLS22
 	 * @exception UnsupportedOperationException if this expression is used with previewEnabled flag as false
 	 * @since 3.39
 	 */
 	public VariableDeclaration getPatternVariable2() {
-		supportedOnlyIn20();
+		supportedOnlyIn22();
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:256]: Requesting for a cleanup here: Please change this name supportedOnlyIn22() to unSupportedBelow22()

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java [hunk 2]
```diff
@@ -276,7 +277,11 @@ boolean subtreeMatch0(ASTMatcher matcher, Object other) {
 	ASTNode clone0(AST target) {
 		TypePattern result = new TypePattern(target);
 		result.setSourceRange(getStartPosition(), getLength());
-		result.setPatternVariable((VariableDeclaration) getPatternVariable().clone(target));
+		if (this.ast.apiLevel < AST.JLS22_INTERNAL) {
+			result.setPatternVariable((VariableDeclaration) getPatternVariable().clone(target));
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:281]: Look at MethodDeclaration.clone0() - see how it has internalSetModifiers() method - this avoids deprecated warnings

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java [hunk 3]
```diff
@@ -276,7 +277,11 @@ boolean subtreeMatch0(ASTMatcher matcher, Object other) {
...
+		} else {
+			result.setPatternVariable((VariableDeclaration) getPatternVariable2().clone(target));
```
Comment (mpalat) [inline org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:283]: cleanup: refer to MethodDeclaration.modifiers - see how it got changed from int to List. Can we think of using a similar approach here

**Review by mpalat — CHANGES_REQUESTED** [review]
Suby,
Let us do some additional cleanup here - Refer to MethodDeclaration - modifiers and see how it got changed. Let us do a similar approach here.

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java (hunk 1 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:256]: I believe the supportedOnlyIn22 method can be handle the logic. Please see the method

```
final void supportedOnlyIn22() {
		if (this.ast.apiLevel < AST.JLS22_INTERNAL) {
			throw new UnsupportedOperationException("Operation only supported in JLS22 AST"); //$NON-NLS-1$
		}
	}
```

Therefore should I create a new method called unSupportedBelow22 ?

Comment (mpalat) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:256]: you can rename the current method if it's used only here

Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:272]: Yes, renaming seems appropriate in this case. However, there are already similar methods like supportedOnlyIn22,  supportedOnlyIn20. etc .. Should I still proceed with renaming the method, or would it be better to keep the existing naming convention for consistency?

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java (hunk 2 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java (hunk 3 above)
Comment (subyssurendran666) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:299]: Done

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java (hunk 1 above)
Comment (mpalat) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:271]: ok. let's not do it for this fix. Please create a new cleanup PR that does this

File: org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java (hunk 3 above)
Comment (mpalat) [reply org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:298]: thanks

**Review by mpalat — APPROVED** [review]
+1
Please rebase and submit. Thanks

**Comment by laeubi** [comment]
@mpalat @subyssurendran666 

There is an javadoc error that fails the build now:

> Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.11.2:javadoc (attach-javadocs) on project org.eclipse.jdt.doc.isv: An error has occurred in Javadoc report generation: 
> Error:  Exit code: 1
> Error:  /home/runner/work/tycho/tycho/aggregator/eclipse.platform.common/bundles/org.eclipse.jdt.doc.isv/../../../eclipse.jdt.core/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:190: error: unknown tag: Deprecated
> Error:  	 * @Deprecated(forRemoval = true, since="2024-07")
> Error:  	   ^

https://github.com/eclipse-jdt/eclipse.jdt.core/blob/5c82c9142f1155b6146c4eb448ae9bea25669252/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java#L190

I assume this should be a method annotation and not part of the javadoc.

Also the date seems wrong `@Deprecated(forRemoval = true, since="2025-03")` is probably wanted?

**Comment by mpalat** [comment]
> @mpalat @subyssurendran666
> 
> There is an javadoc error that fails the build now:
> 
> ```
> Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.11.2:javadoc (attach-javadocs) on project org.eclipse.jdt.doc.isv: An error has occurred in Javadoc report generation: 
> Error:  Exit code: 1
> Error:  /home/runner/work/tycho/tycho/aggregator/eclipse.platform.common/bundles/org.eclipse.jdt.doc.isv/../../../eclipse.jdt.core/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:190: error: unknown tag: Deprecated
> Error:  	 * @Deprecated(forRemoval = true, since="2024-07")

Thanks @laeubi  for flagging this.
@subyssurendran666 : "D" in deprecated should be small "d" - could you please correct the same?

**Comment by subyssurendran666** [comment]
> > @mpalat @subyssurendran666
> > There is an javadoc error that fails the build now:
> > ```
> > Error:  Failed to execute goal org.apache.maven.plugins:maven-javadoc-plugin:3.11.2:javadoc (attach-javadocs) on project org.eclipse.jdt.doc.isv: An error has occurred in Javadoc report generation: 
> > Error:  Exit code: 1
> > Error:  /home/runner/work/tycho/tycho/aggregator/eclipse.platform.common/bundles/org.eclipse.jdt.doc.isv/../../../eclipse.jdt.core/org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/TypePattern.java:190: error: unknown tag: Deprecated
> > Error:  	 * @Deprecated(forRemoval = true, since="2024-07")
> > ```
> 
> Thanks @laeubi for flagging this. @subyssurendran666 : "D" in deprecated should be small "d" - could you please correct the same?

Hello @mpalat @laeubi,

I was referring this [doc](https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/Eclipse_API_Central_Deprecation_Policy.md#identifying-deprecated-api) for the deprecation and it is clearly mentioned that, the javaDoc changes should be `@Deprecated(forRemoval = true, since="2024-12")`. That is why the deprecated in the javadoc written as `D`. I will modify that with `d(lower case)`.

The reason why I have added the since="2024-07", because this should be deprecated since 7th month of 2024. However after referring @laeubi comment, I got idea that this should be `2025-03`.

**Comment by laeubi** [comment]
> I was referring this [doc](https://github.com/eclipse-platform/eclipse.platform/blob/master/docs/Eclipse_API_Central_Deprecation_Policy.md#identifying-deprecated-api) for the deprecation and it is clearly mentioned that, the javaDoc changes should be @Deprecated(forRemoval = true, since="2024-12").

Please read again! As also mentioned above this is **not an javadoc tag** it is a method annotation as also mentioned in the doc:

> Java API is deprecated through use of the `@Deprecate` **annotation**.

there is also a javadoc tag `@deprecated` but it has different purpose, see for example:

https://stackoverflow.com/questions/20674135/deprecated-vs-deprecated

**Comment by iloveeclipse** [comment]
@laeubi :  https://github.com/eclipse-jdt/eclipse.jdt.core/pull/3504

**Comment by jukzi** [comment]
this PR caused regression in ui: https://github.com/eclipse-jdt/eclipse.jdt.ui/issues/1899

**Comment by rgrunber** [comment]
I've also noticed a regression (initially from the JDT-LS side). Affecting "Pattern matching for instanceof" cleanup.

https://github.com/user-attachments/assets/ae08f056-913f-4937-aefd-742d398c8e29

**Comment by mpalat** [comment]
@subyssurendran666 Given that there is regression, can you please revert this PR for now?

**Pull request merged**